processing/src/main/java/org/apache/druid/query/scan/ScanQueryRunnerFactory.java

```diff
@@ -275,6 +367,63 @@ Sequence<ScanResultValue> stableLimitingSort(
     }
   }
 
+  Sequence<ScanResultValue> multiColumnSort(
+      Sequence<ScanResultValue> inputSequence,
+      ScanQuery scanQuery
+  ) throws IOException
+  {
+
+    if (scanQuery.getScanRowsLimit() > Integer.MAX_VALUE) {
+      throw new UOE(
+          "Limit of %,d rows not supported for priority queue strategy of non-time-ordering scan results",
+          scanQuery.getScanRowsLimit()
+      );
+    }
+    // Converting the limit from long to int could theoretically throw an ArithmeticException but this branch
+    // only runs if limit < MAX_LIMIT_FOR_IN_MEMORY_TIME_ORDERING (which should be < Integer.MAX_VALUE)
+    int limit = Math.toIntExact(scanQuery.getScanRowsLimit());
+    List<String> sortColumns = scanQuery.getOrderBys().stream().map(orderBy -> orderBy.getColumnName()).collect(Collectors.toList());
+    MultiColumnSorter<ScanResultValue> multiColumnSorter = new QueueBasedMultiColumnSorter<>(limit, scanQuery.getOrderByNoneTimeResultOrdering());
+    Yielder<ScanResultValue> yielder = Yielders.each(inputSequence);
+    try {
+      boolean doneScanning = yielder.isDone();
+      // We need to scan limit elements and anything else in the last segment
+      while (!doneScanning) {
+        ScanResultValue next = yielder.get();
+        List<ScanResultValue> singleEventScanResultValues = next.toSingleEventScanResultValues();
+        for (ScanResultValue srv : singleEventScanResultValues) {
+          // Using an intermediate unbatched ScanResultValue is not that great memory-wise, but the column list
+          // needs to be preserved for queries using the compactedList result format
+          List<Integer> idxs = sortColumns.stream().map(c -> srv.getColumns().indexOf(c)).collect(Collectors.toList());
```

Comment:
I think this may be a bit inefficient, and may even be unnecessary. The rows are already in index order. Looking for row c at position x will return x. The index can be built simply as looping for 0 to n and inserting i at each position. Or an I missing something?

We're doing the clever thing and sorting a list of integer indexes, rather than moving rows. But, note: the list is boxed integer: an object. The items in the "main list" are also objects. So, it is actually cheaper to just sort the objects in the main list rather than the index list. Reason: in either case, we're just swapping pointers.

If, however, the data were in a buffer, and we had to physically move bytes (rather than pointers), then the indirection would be a nice performance improvement.

If we did use an indirection, then using an array of (unboxed) ints can be more memory-friendly.